src/cls/cmpxattr/client.h

```diff
@@ -0,0 +1,64 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab ft=cpp
+
+/*
+ * Ceph - scalable distributed file system
+ *
+ * Copyright (C) 2024 IBM Corp.
+ *
+ * This is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License version 2.1, as published by the Free Software
+ * Foundation.  See file COPYING.
+ */
+
+#pragma once
+
+#include <optional>
+#include "include/rados/librados_fwd.hpp"
+#include "types.h"
+#include "ops.h"
+#include "include/utime.h"
+#include "common/ceph_time.h"
+#include "common/dout.h"
+
+namespace cls::cmpxattr {
+  /// requests with too many key comparisons will be rejected with -E2BIG
+  static constexpr uint32_t max_keys = 8;
+
+  /// process each of the xattrs value comparisons according to the same rules as
+  /// cmpxattr(). IFF **all** key/value pairs in @cmp_pairs compare successfully
+  /// a set operation will be perfrom for **all** key/value pairs in @set_pairs.
+  /// Caller must supply non-empty cmp_pairs and set_pairs
+  /// However, it is legal to pass an empty value bl for any key
+  ///         An empty value will match key with an empty value or a non-existing key!
+  ///
+  /// for comparisons with Mode::U64, failure to decode an input value is
+  /// reported as -EINVAL.
+  /// a decode failure of a stored value is treated as an unsuccessful comparison
+  /// and is not reported as an error
+  [[nodiscard]] int cmp_vals_set_vals(librados::ObjectWriteOperation& writeop,
+				      Mode mode, Op comparison,
+				      const ComparisonMap&& cmp_pairs,
+				      const std::map<std::string, bufferlist>&& set_pairs,
```

Comment:
these should either be `const&` for copy or non-const `&&` for move - you can't move from a `const&&`

Reply:
I need to be able to compare on some attributes (ETAG, size) and set other attributes (manifest etc.)
I also need to set new attributes while doing this (marking the object as a source of dedup)

Reply:
can't your transaction start with a `ObjectOperation::cmpxattr()` followed by a series or `setxattr()`s?